g3/git/shell.py

```diff
@@ -0,0 +1,237 @@
+            stdin_writer.close()
+
+        async def run() -> Tuple[int, bytes, bytes]:
+            proc = await asyncio.create_subprocess_exec(
+                *args,
+                stdin=stdin,
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+                cwd=self.cwd,
+                env=env,
+            )
+            proc_stdout = proc.stdout
+            proc_stderr = proc.stderr
+            assert proc_stdout is not None
+            assert proc_stderr is not None
+            _, out, err, _ = await asyncio.gather(
+                feed_input(proc.stdin),
+                process_stream(proc_stdout, stdout, sys.stdout),
+                process_stream(proc_stderr, stderr, sys.stderr),
+                proc.wait(),
+            )
+            assert proc.returncode is not None
+            return proc.returncode, out, err
+
+        loop = asyncio.get_event_loop()
+        returncode, out, err = loop.run_until_complete(run())
+
+        if err:
+            logging.debug("# stderr:\n" + err.decode(errors="backslashreplace"))
+        if out:
+            logging.debug(("# stdout:\n" if err else "") + out.decode(errors="backslashreplace").replace("\0", "\\0"))
+
+        if exitcode:
+            logging.debug("Exit code: {}".format(returncode))
+            return returncode == 0
+        if returncode != 0:
+            raise RuntimeError("{} failed with exit code {}. Output was '{}'".format(" ".join(args), returncode, out))
+
+        if stdout == subprocess.PIPE:
+            return out.decode()  # do a strict decode for actual return
+        else:
+            return None
+
+    def _maybe_rstrip(self, s: _SHELL_RET) -> _SHELL_RET:
+        if isinstance(s, str):
+            return s.rstrip()
+        else:
+            return s
+
+    @overload  # noqa: F811
+    def git(self, *args: str) -> str:
+        ...
+
+    @overload  # noqa: F811
+    def git(self, *args: str, input: str) -> str:
+        ...
+
+    @overload  # noqa: F811
+    def git(self, *args: str, **kwargs: Any) -> _SHELL_RET:
+        ...
+
+    def git(self, *args: str, **kwargs: Any) -> _SHELL_RET:  # noqa: F811
+        """
+        Run a git command.  The returned stdout has trailing newlines stripped.
+
+        Args:
+            *args: Arguments to git
+            **kwargs: Any valid kwargs for sh()
+        """
+        env = kwargs.setdefault("env", {})
+        if self.testing:
+            env.setdefault("EDITOR", ":")
+            env.setdefault("GIT_MERGE_AUTOEDIT", "no")
+            env.setdefault("LANG", "C")
+            env.setdefault("LC_ALL", "C")
+            env.setdefault("PAGER", "cat")
+            env.setdefault("TZ", "UTC")
+            env.setdefault("TERM", "dumb")
+            if "stderr" not in kwargs:
+                kwargs["stderr"] = subprocess.PIPE
+
+        return self._maybe_rstrip(self.sh(*(("git",) + args), **kwargs))
+
+    def repo_name(self):  # noqa: F811
```

Comment:
```suggestion
   @property
   def repo_name(self):
```
We can transform this method to a property so that we can access it without `()`:
```
shell.repo_name
```
instead of 
```
shell.repo_name()
```